Sources/Lifecycle/Lifecycle.swift

```diff
@@ -247,17 +248,28 @@ public struct ServiceLifecycle {
         }
     }
 
-    private func setupShutdownHook() {
+    private func installSignalhandler() {
+        if self.configuration.shutdownSignal != nil, ServiceLifecycle.signalHandlerInstalled.compareAndSwap(expected: false, desired: true) {
+            self.register(label: "Shutdown hooks",
+                          start: .sync(self.installShutdownHook),
+                          shutdown: .sync(self.deinstallShutdownHook))
+        }
+    }
+
+    private func installShutdownHook() {
         self.configuration.shutdownSignal?.forEach { signal in
-            self.log("setting up shutdown hook on \(signal)")
-            let signalSource = ServiceLifecycle.trap(signal: signal, handler: { signal in
+            ServiceLifecycle.signalHandlerSources.append(ServiceLifecycle.trap(signal: signal, handler: { signal in
                 self.log("intercepted signal: \(signal)")
                 self.shutdown()
-            }, cancelAfterTrap: true)
-            self.underlying.shutdownGroup.notify(queue: .global()) {
-                signalSource.cancel()
-            }
+            }, cancelAfterTrap: true))
+        }
+    }
+
+    private func deinstallShutdownHook() {
+        for source in ServiceLifecycle.signalHandlerSources {
+            source.cancel()
         }
+        ServiceLifecycle.signalHandlerSources = []
     }
```

Comment:
see above, not sure why this is required

Reply:
It was intended to release any references to the signal dispatch sources. Maybe it was a bit overkill and I can revert to the previous cancellation approach which would only disable the signal dispatch source for the signal that was triggered.